File: src/mediathek/daten/ListePsetVorlagen.java
/*    
 *    MediathekView
 *    Copyright (C) 2008   W. Xaver
 *    W.Xaver[at]googlemail.com
 *    http://zdfmediathk.sourceforge.net/
 *    
 *    This program is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package mediathek.daten;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;
import java.util.LinkedList;
import javax.swing.JFrame;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLStreamConstants;
import javax.xml.stream.XMLStreamReader;
import mediathek.controller.IoXmlLesen;
import mediathek.controller.Log;
import mediathek.file.GetFile;
import mediathek.tool.Funktionen;
import mediathek.tool.Konstanten;
import mediathek.tool.TModel;
import msearch.tool.MSConst;

public class ListePsetVorlagen extends LinkedList<String[]> {

    public static final String BS_WIN_32 = "Windows-32Bit";
    public static final String BS_WIN_64 = "Windows-64Bit";
    public static final String BS_LINUX = "Linux";
    public static final String BS_MAC = "Mac";
    public static final String[] BS = {"", BS_WIN_32, BS_WIN_64, BS_LINUX, BS_MAC};
    //
    public static final String PGR = "Vorlage";
    public static final String PGR_NAME = "Name";
    public static final int PGR_NAME_NR = 0;
    public static final String PGR_BESCHREIBUNG = "Beschreibung";
    public static final int PGR_BESCHREIBUNG_NR = 1;
    public static final String PGR_VERSION = "Version";
    public static final int PGR_VERSION_NR = 2;
    public static final String PGR_BS = "Bs";
    public static final int PGR_BS_NR = 3;
    public static final String PGR_URL = "URL";
    public static final int PGR_URL_NR = 4;
    public static final String PGR_INFO = "Info";
    public static final int PGR_INFO_NR = 5;
    public static final int PGR_MAX_ELEM = 6;
    public static final String[] PGR_COLUMN_NAMES = {PGR_NAME, PGR_BESCHREIBUNG, PGR_VERSION, PGR_BS, PGR_URL, PGR_INFO};
    private final static int TIMEOUT = 10000;

    public TModel getTModel(String bs) {
        LinkedList<String[]> tmp = new LinkedList<>();
        String[][] object;
        if (this.size() > 0) {
            if (!bs.equals("")) {
                for (String[] aThi : this) {
                    if (aThi[PGR_BS_NR].contains(bs)) {
                        tmp.add(aThi);
                    }
                }
                object = new String[tmp.size()][PGR_MAX_ELEM];
                for (int i = 0; i < tmp.size(); i++) {
                    object[i] = tmp.get(i);
                }
            } else {
                object = new String[this.size()][PGR_MAX_ELEM];
                for (int i = 0; i < this.size(); i++) {
                    object[i] = this.get(i);
                }
            }
            return new TModel(object, PGR_COLUMN_NAMES);
        } else {
            return new TModel(new Object[][]{}, PGR_COLUMN_NAMES);
        }
    }

    public boolean getListe() {
        try {
            this.clear();
            int event;
            XMLInputFactory inFactory = XMLInputFactory.newInstance();
            inFactory.setProperty(XMLInputFactory.IS_COALESCING, Boolean.FALSE);
            XMLStreamReader parser;
            InputStreamReader inReader;
            URLConnection conn;
            conn = new URL(Konstanten.ADRESSE_VORLAGE_PROGRAMMGRUPPEN).openConnection();
            conn.setRequestProperty("User-Agent", Daten.getUserAgent());
            conn.setReadTimeout(TIMEOUT);
            conn.setConnectTimeout(TIMEOUT);
            inReader = new InputStreamReader(conn.getInputStream(), MSConst.KODIERUNG_UTF);
            parser = inFactory.createXMLStreamReader(inReader);
            while (parser.hasNext()) {
                event = parser.next();
                if (event == XMLStreamConstants.START_ELEMENT) {
                    if (parser.getLocalName().equals(PGR)) {
                        //wieder ein neuer Server, toll
                        String[] p = new String[PGR_MAX_ELEM];
                        get(parser, PGR, PGR_COLUMN_NAMES, p);
                        if (!p[PGR_URL_NR].equals("")) {
                            this.add(p);
                        }
                    }
                }
            }
        } catch (Exception ex) {
            Log.fehlerMeldung(398001963,  "VorlageProgrammgruppen.getListe", ex);
            return false;
        }
        return true;
    }

    public static ListePset getStandarset(JFrame parent, Daten ddaten) {
        ListePset pSet = null;
        String[] vorlage = null;
        ListePsetVorlagen lv = new ListePsetVorlagen();
        if (lv.getListe()) {
            for (String[] ar : lv) {
                if (ar[PGR_NAME_NR].equalsIgnoreCase("Standardset " + Funktionen.getOsString())) {
                    vorlage = ar;
                    break;
                }
            }
            if (vorlage != null) {
                if (!vorlage[PGR_URL_NR].equals("")) {
                    pSet = IoXmlLesen.importPset(parent, ddaten, vorlage[ListePsetVorlagen.PGR_URL_NR], true);
                    if (pSet != null) {
                        pSet.version = vorlage[PGR_VERSION_NR];
                    }
                }
            }
        }
        if (pSet == null) {
            // dann nehmen wir halt die im jar-File
            pSet = getStandardprogramme(parent, ddaten);
        }
        return pSet;
    }

    private static ListePset getStandardprogramme(JFrame parent, Daten ddaten) {
        // liefert das Standard Programmset für das entsprechende BS
        ListePset pSet;
        InputStream datei;
        switch (Funktionen.getOs()) {
            case LINUX:
                datei = new GetFile().getPsetVorlageLinux();
                break;
            case MAC:
                datei = new GetFile().getPsetVorlageMac();
                break;

            default:
                datei = new GetFile().getPsetVorlageWindows();
        }
        // Standardgruppen laden
        pSet = IoXmlLesen.importPset(parent, ddaten, datei, true);
        return pSet;
    }

    private boolean get(XMLStreamReader parser, String xmlElem, String[] xmlNames, String[] strRet) {
        boolean ret = true;
        int maxElem = strRet.length;
        for (int i = 0; i < maxElem; ++i) {
            strRet[i] = "";
        }
        try {
            while (parser.hasNext()) {
                int event = parser.next();
                if (event == XMLStreamConstants.END_ELEMENT) {
                    if (parser.getLocalName().equals(xmlElem)) {
                        break;
                    }
                }
                if (event == XMLStreamConstants.START_ELEMENT) {
                    for (int i = 0; i < maxElem; ++i) {
                        if (parser.getLocalName().equals(xmlNames[i])) {
                            strRet[i] = parser.getElementText();
                            break;
                        }
                    }
                }
            }
        } catch (Exception ex) {
            ret = false;
            Log.fehlerMeldung(467256394,  "VorlageProgrammgruppen.get", ex);
        }
        return ret;
    }
}

File: src/main/kotlin/mediathek/daten/ListePsetVorlagen.kt
package mediathek.daten

import mediathek.config.Konstanten
import mediathek.tool.NetUtils
import mediathek.tool.http.MVHttpClient
import mediathek.tool.models.NonEditableTableModel
import okhttp3.Request
import org.apache.commons.lang3.SystemUtils
import org.apache.logging.log4j.LogManager
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.io.StringReader
import java.net.ConnectException
import java.net.UnknownHostException
import java.nio.charset.StandardCharsets
import javax.swing.JFrame
import javax.swing.table.TableModel
import javax.xml.stream.XMLInputFactory
import javax.xml.stream.XMLStreamConstants
import javax.xml.stream.XMLStreamException
import javax.xml.stream.XMLStreamReader

class ListePsetVorlagen : ArrayList<Array<String>>() {
    fun createModel(bs: String): TableModel {
        if (isEmpty()) {
            return NonEditableTableModel(emptyArray(), programGroupColumnNames())
        }

        val rows = if (bs.isNotEmpty()) {
            filter { row -> row[PGR_BS_NR].contains(bs) }
        } else {
            this
        }
        val data = Array(rows.size) { index ->
            Array<Any?>(PGR_MAX_ELEM) { column -> rows[index][column] }
        }
        return NonEditableTableModel(data, programGroupColumnNames())
    }

    fun loadListOfSets(): Boolean {
        try {
            clear()

            val inFactory = xmlInputFactory()
            val url = requireNotNull(Konstanten.URL_MEDIATHEKVIEW_RESOURCES.resolve(Konstanten.PSET_PROGRAM_GROUP_LIST_PATH))
            val request = Request.Builder().url(url).get().build()
            MVHttpClient.httpClient.newCall(request).execute().use { response ->
                response.body.use { body ->
                    if (!response.isSuccessful) {
                        return false
                    }

                    body.byteStream().use { input ->
                        InputStreamReader(input, StandardCharsets.UTF_8).use { reader ->
                            inFactory.createXMLStreamReader(reader).use { parser ->
                                while (parser.hasNext()) {
                                    if (parser.next() == XMLStreamConstants.START_ELEMENT && parser.localName == PGR) {
                                        val programGroup = Array(PGR_MAX_ELEM) { "" }
                                        get(parser, PGR, PGR_COLUMN_NAMES, programGroup)
                                        if (programGroup[PGR_URL_NR].isNotEmpty()) {
                                            add(programGroup)
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        } catch (_: UnknownHostException) {
            return false
        } catch (_: ConnectException) {
            return false
        } catch (ex: Exception) {
            logger.error("loadListOfSets()", ex)
            return false
        }

        return true
    }

    companion object {
        private const val BS_WIN_64 = "Windows-64Bit"
        private const val BS_LINUX = "Linux"
        private const val BS_MAC = "Mac"

        val BS: Array<String> = arrayOf("", BS_WIN_64, BS_LINUX, BS_MAC)

        const val PGR = "Vorlage"
        const val PGR_NAME = "Name"
        const val PGR_NAME_NR = 0
        const val PGR_BESCHREIBUNG = "Beschreibung"
        const val PGR_BESCHREIBUNG_NR = 1
        const val PGR_VERSION = "Version"
        const val PGR_VERSION_NR = 2
        const val PGR_BS = "Bs"
        const val PGR_BS_NR = 3
        const val PGR_URL = "URL"
        const val PGR_URL_NR = 4
        const val PGR_INFO = "Info"
        const val PGR_INFO_NR = 5
        const val PGR_MAX_ELEM = 6

        val PGR_COLUMN_NAMES: Array<String> = arrayOf(PGR_NAME, PGR_BESCHREIBUNG, PGR_VERSION, PGR_BS, PGR_URL, PGR_INFO)

        private val logger = LogManager.getLogger(ListePsetVorlagen::class.java)

        fun getStandarset(parent: JFrame?, replaceMuster: Boolean): ListePset? {
            var listePset: ListePset? = null
            var vorlage: Array<String>? = null
            val listePsetVorlagen = ListePsetVorlagen()
            if (listePsetVorlagen.loadListOfSets()) {
                for (row in listePsetVorlagen) {
                    if (row[PGR_NAME_NR].equals("Standardset ${getOperatingSystemString()}", ignoreCase = true)) {
                        vorlage = row
                        break
                    }
                }
                if (vorlage != null && vorlage[PGR_URL_NR].isNotEmpty()) {
                    listePset = importPsetFile(vorlage[PGR_URL_NR], true)
                    if (listePset != null) {
                        listePset.version = vorlage[PGR_VERSION_NR]
                    }
                }
            }
            if (listePset == null) {
                // dann nehmen wir halt die im jar-File
                // liefert das Standard Programmset für das entsprechende BS
                // Standardgruppen laden
                listePset = importPset(getLocalPsetTemplate(), true)
            }

            if (replaceMuster && listePset != null) {
                // damit die Variablen ersetzt werden
                ProgramSetTemplateResolver.replaceTemplates(parent, listePset)
            }
            return listePset
        }

        fun importPsetFile(dateiUrl: String, log: Boolean): ListePset? =
            try {
                if (NetUtils.isUrl(dateiUrl)) {
                    val request = Request.Builder().url(dateiUrl).get().build()
                    MVHttpClient.httpClient.newCall(request).execute().use { response ->
                        response.body.use { body ->
                            if (response.isSuccessful) {
                                body.byteStream().use { input ->
                                    InputStreamReader(input, StandardCharsets.UTF_8).use { reader ->
                                        importPset(reader, log)
                                    }
                                }
                            } else {
                                null
                            }
                        }
                    }
                } else {
                    FileInputStream(dateiUrl).use { input ->
                        InputStreamReader(input, StandardCharsets.UTF_8).use { reader ->
                            importPset(reader, log)
                        }
                    }
                }
            } catch (ex: Exception) {
                if (log) {
                    logger.error("importPsetFile(..)", ex)
                }
                null
            }

        fun importPsetText(text: String, log: Boolean): ListePset? =
            try {
                importPset(StringReader(text), log)
            } catch (_: IOException) {
                null
            }

        private fun getOperatingSystemString(): String =
            when {
                SystemUtils.IS_OS_MAC_OSX -> "Mac"
                SystemUtils.IS_OS_WINDOWS -> "Windows"
                SystemUtils.IS_OS_LINUX -> "Linux"
                else -> ""
            }

        private fun getProgramSetTemplateFromLocalResources(): String =
            when {
                SystemUtils.IS_OS_LINUX -> "/mediathek/file/pset_linux.xml"
                SystemUtils.IS_OS_MAC_OSX -> "/mediathek/file/pset_mac.xml"
                SystemUtils.IS_OS_WINDOWS -> "/mediathek/file/pset_windows.xml"
                else -> error("Unsupported OS")
            }

        private fun getLocalPsetTemplate(): InputStreamReader? =
            try {
                val path = getProgramSetTemplateFromLocalResources()
                val stream = requireNotNull(ListePsetVorlagen::class.java.getResource(path)).openStream()
                InputStreamReader(stream, StandardCharsets.UTF_8)
            } catch (ex: IOException) {
                logger.error("getLocalPsetTemplate()", ex)
                null
            }

        private fun importPset(reader: java.io.Reader?, log: Boolean): ListePset? {
            if (reader == null) {
                return null
            }

            var datenPset: DatenPset? = null
            val liste = ListePset()

            try {
                xmlInputFactory().createXMLStreamReader(reader).use { parser ->
                    while (parser.hasNext()) {
                        if (parser.next() == XMLStreamConstants.START_ELEMENT) {
                            when (parser.localName) {
                                DatenPset.TAG -> {
                                    datenPset = DatenPset()
                                    val psetValues = Array(DatenPset.MAX_ELEM) { "" }
                                    if (!get(parser, DatenPset.TAG, DatenPset.XML_NAMES, psetValues)) {
                                        datenPset = null
                                    } else {
                                        datenPset.copyFrom(psetValues)
                                        if (!datenPset.isEmpty()) {
                                            // kann beim Einlesen der Konfigdatei vorkommen
                                            liste.add(datenPset)
                                        }
                                    }
                                }

                                DatenProg.TAG -> {
                                    val currentPset = datenPset
                                    if (currentPset != null) {
                                        val datenProg = DatenProg()
                                        val progValues = datenProg.toArray()
                                        if (get(parser, DatenProg.TAG, DatenProg.XML_NAMES, progValues)) {
                                            datenProg.copyFrom(progValues)
                                            currentPset.addProg(datenProg)
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            } catch (ex: Exception) {
                if (log) {
                    logger.error("importPset", ex)
                }
                return null
            } finally {
                try {
                    reader.close()
                } catch (_: IOException) {
                }
            }

            return liste.ifEmpty { null }
        }

        private fun get(
            parser: XMLStreamReader,
            xmlElem: String,
            xmlNames: Array<String>,
            result: Array<String>,
        ): Boolean {
            result.fill("")
            return try {
                while (parser.hasNext()) {
                    val event = parser.next()
                    if (event == XMLStreamConstants.END_ELEMENT && parser.localName == xmlElem) {
                        break
                    }
                    if (event == XMLStreamConstants.START_ELEMENT) {
                        val index = xmlNames.indexOf(parser.localName)
                        if (index != -1 && index < result.size) {
                            result[index] = parser.elementText
                        }
                    }
                }
                true
            } catch (ex: Exception) {
                logger.error("get", ex)
                false
            }
        }

        private fun xmlInputFactory(): XMLInputFactory =
            XMLInputFactory.newInstance().apply {
                setProperty(XMLInputFactory.IS_COALESCING, false)
                setProperty(XMLInputFactory.SUPPORT_DTD, false)
                setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false)
            }

        private fun programGroupColumnNames(): Array<Any?> =
            Array(PGR_COLUMN_NAMES.size) { index -> PGR_COLUMN_NAMES[index] }

        private inline fun XMLStreamReader.use(block: (XMLStreamReader) -> Unit) {
            try {
                block(this)
            } finally {
                try {
                    close()
                } catch (_: XMLStreamException) {
                }
            }
        }
    }
}
